### ISSUE TITLE
Eliminate return value on Customize

### ISSUE THREAD

**Issue by cocogorilla** [AutoFixture/AutoFixture#977]
Referencing stack thread: [https://stackoverflow.com/questions/48672741/customize-generic-vs-non-generic-return-types?noredirect=1#comment84344170_48672741](api confusion)

I've found myself confused by the API for Customize overloads, whether or not the fixture is being mutated, enough times I finally decided to ask.

Mark suggests, since the internals ARE mutating the fixture, that the return value be eliminated in keeping with CQS.

**Comment by zvirja** [comment]
@cocogorilla Thanks for raising the issue here! Actually, a good point 👍

Currently we have 3 very related APIs today:
```c#
IFixture Customize(ICustomization customization);
void Customize<T>(Func<ICustomizationComposer<T>, ISpecimenBuilder> composerTransformation);
ICustomizationComposer<T> Build<T>();
```

The API description is following:
- the first one was designed to suppose the Fluent API (so you could apply a few `ICustomization` instances in a row). It _mutates_ the fixture instance.
- the second one allows to apply customization for the specific type in a fluent matter (you chain the `composerTransformation` methods). It _mutates_ the fixture instance.
- the third one is to perform the one-time customization and build the instance of the configured type. It also provides fluent interface (you should use the returned value). This method doesn't mutate the fixture itself and return the modified builder.

If we analyze that from [CommandQuerySeparation](https://martinfowler.com/bliki/CommandQuerySeparation.html) perspective, we have violation for the following signature only:
```c#
IFixture Customize(ICustomization customization);
```

Likely, I agree with the idea mentioned on the StackOverflow, so we should return the void value here (as we are not mutating the fixture). Instead, to simplify the API of applying multiple customizations together, I suggest to change the overload to take the variable number of customizations (i.e. `params` keyword):
```c#
fixture.Customize(new Customization1(),
                  new OtherCustomization(),
                  new ThirdCustomization());
```

That is the breaking change, so we could release it in v5 only.

@moodmosaic Do you have some other vision regarding the situation? 😉

**Comment by moodmosaic** [comment]
```c#
IFixture Customize(ICustomization customization);
```

can be changed to

```c#
void Customize(ICustomization customization);
```

and then, you can pass a [CompositeCustomization](https://github.com/AutoFixture/AutoFixture/blob/master/Src/AutoFixture/CompositeCustomization.cs) if you need to customize the `IFixture` instance using multiple customizations.

---

See also: https://stackoverflow.com/a/48673138

**Comment by zvirja** [comment]
> can be changed to `void Customize(ICustomization customization);`

Yep, I know about `CompositeCustomization`, however it looks a bit clumsy and verbose to me:

```c#
var fixture = new Fixture();
fixture.Customize(
    new AutoNSubstituteCustomization(),
    new NoAutoPropertiesCustomization(),
    new DummyCustomization(),
    new NoDataAnnotationsCustomization());
```
vs
```c#
var fixture = new Fixture();
fixture.Customize(new CompositeCustomization(
    new AutoNSubstituteCustomization(),
    new NoAutoPropertiesCustomization(),
    new DummyCustomization(),
    new NoDataAnnotationsCustomization()));
```

The first option looks more lightweight and simpler to me. Also it's more user friendly, as people might know nothing about `CompositeCustomization`.

As for the `CompositeCustomization`, I see its value when you would like to enclose a few customizations into a single package, so you derive from `CompositeCustomization` and pass a few customizations to the parent constructor.

----

The point is that of course, we can live without `params` API, however it provides much simpler experience and I'm fully for it.

**Comment by Kralizek** [comment]
You can also create an extension method that accepts a set of customizations and internally just forwards a CompositeCustomization.

I usually use this approach when I want to keep the API of a certain service clean but still offer convenience methods for those who want.

Something as easy as
```
public static class Exts
{
    public static void Customize(this IFixture fixture, params ICustomization[] customizations)
    {
        fixture.Customize(new CompositeCustomization(customizations));
    }
}
```

**Comment by zvirja** [comment]
> I usually use this approach when I want to keep the API of a certain service clean but still offer convenience methods for those who want.

This includes a bit different story - we should move the [`Customize(ICustomization c)` method](https://github.com/AutoFixture/AutoFixture/blob/45832a734087556fa9934a62babd75918b468a6a/Src/AutoFixture/IFixture.cs#L103) out of the `IFixture` interface and make it an extension method. I don't see any reason why `Fixture` should implement this method, given that no internal stuff is required.

I do understand that user could create its own `Customize()` extension method consuming the `params`, but I don't see any reasons we cannot provide that convenient API out of the box, given the specifics of this method (it's often used to apply multiple customizations).

**Comment by moodmosaic** [comment]
What @Kralizek proposes

    public static void Customize(this IFixture fixture, params ICustomization[] customizations)
    {
        fixture.Customize(new CompositeCustomization(customizations));
    }

may also be declared externally; AutoFixture doesn't have to support each and every usage scenario.

---

The idea is that AutoFixture provides a decent foundation of extensible blocks; it doesn't have to be a monolith. If you feel like adding this into AutoFixture might help most of its users, sure, go for it.

**Comment by Kralizek** [comment]
It has to be said that I'd rather have a "author curated" set of extensions, say a "AutoFixture.Extensions" package, rather than having to reinvent the wheel myself.

I am already bothered by having to create myself a package to put together Moq and AutoData attributes from the NUnit3 glue assembly :)

**Comment by moodmosaic** [comment]
>I'd rather have a "author curated" set of extensions, say a "AutoFixture.Extensions"

No, that's too much of maintenance 😅 It'll either end up being in the core library or you do it on your own in your codebase(s). After all, you *do* need to define some ad-hoc customization(s) anyway...

**Comment by zvirja** [comment]
> No, that's too much of maintenance 

Agree, for this particular case it would be an overkill ☺️

 > If you feel like adding this into AutoFixture might help most of its users, sure, go for it.

Well, yes, I think that users will benefit from `params`, as API is more friendly out-of-the-box. From the architectural perspective, I believe it doesn't matter too much for us, whether we accept single customization or a list of them.

**Comment by aivascu** [comment]
I know this is an old one but since it was planned for `v5` I feel like have to explain my decision for future reference.
I don't think removing the `IFixture` return type would be a good decision. At the moment most `AutoDataAttributes` accept as a the first argument a `Func<IFixture>`. This means that most of them are declared like this.
```cs
class MyScenarioDataAttribute : AutoDataAttribute
{
    public MyScenarioDataAttribute()
        : base(() => new Fixture().Customize(new MyScenarioCustomization()))
    { }
}
```
Removing the return type would mean that users would have to declare factory functions or implement their own extension methods to compensate for the removal of the return type. Not only this will make the code more verbose, it would also be a pain to deal with in existing code-bases.

On the other hand adding the `IFixture` return type to `Customize<T>()` would make much more sense since it would allow chaining type customizations. But this is an idea for future versions.

```cs
var fixture = new Fixture()
    .Customize(new MyScenarioCustomization())
    .Customize<Product>(c => c.With(x => x.Price, 100))
    .Customize<Customer>(c => c.With(x => x.Credit, 1000));
```
